s3transfer/manager.py

```diff
@@ -0,0 +1,165 @@
+        """
+        self.multipart_threshold = multipart_threshold
+        self.max_concurrency = max_concurrency
+        self.multipart_chunksize = multipart_chunksize
+        self.max_queue_size = max_queue_size
+
+
+class TransferManager(object):
+    VALID_SUBSCRIBER_TYPES = [
+        'queued',
+        'progress',
+        'done'
+    ]
+
+    ALLOWED_UPLOAD_ARGS = [
+        'ACL',
+        'CacheControl',
+        'ContentDisposition',
+        'ContentEncoding',
+        'ContentLanguage',
+        'ContentType',
+        'Expires',
+        'GrantFullControl',
+        'GrantRead',
+        'GrantReadACP',
+        'GrantWriteACL',
+        'Metadata',
+        'RequestPayer',
+        'ServerSideEncryption',
+        'StorageClass',
+        'SSECustomerAlgorithm',
+        'SSECustomerKey',
+        'SSECustomerKeyMD5',
+        'SSEKMSKeyId',
+    ]
+
+    def __init__(self, client, config=None):
+        """A transfer manager interface for Amazon S3
+
+        :param client: Client to be used by the manager
+        :param config: TransferConfig to associate specific configurations
+        """
+        self._client = client
+        self._config = config
+        if config is None:
+            self._config = TransferConfig()
+        self._osutil = OSUtils()
+        self._executor = BoundedExecutor(
+            max_size=self._config.max_queue_size,
+            max_num_threads=self._config.max_concurrency
+        )
+
+    def upload(self, fileobj, bucket, key, extra_args=None, subscribers=None):
+        """Uploads a file to S3
+
+        :type fileobj: str
+        :param fileobj: The name of a file to upload.
+
+        :type bucket: str
+        :param bucket: The name of the bucket to upload to
+
+        :type key: str
+        :param key: The name of the key to upload to
+
+        :type extra_args: dict
+        :param extra_args: Extra arguments that may be passed to the
+            client operation
+
+        :type subscribers: a list of subscribers
+        :param subscribers: The list of subscribers to be invoked in the
+            order provided based on the event emit during the process of
+            the transfer request.
+
+        :rtype: s3transfer.futures.TransferFuture
+        :returns: Transfer future representing the upload
+        """
+        if extra_args is None:
+            extra_args = {}
+        if subscribers is None:
+            subscribers = []
+        self._validate_subscribers(subscribers)
+        self._validate_all_known_args(extra_args, self.ALLOWED_UPLOAD_ARGS)
+        call_args = CallArgs(
+            fileobj=fileobj, bucket=bucket, key=key, extra_args=extra_args,
+            subscribers=subscribers
+        )
+        upload_submitter = UploadTaskSubmitter(
+            client=self._client, osutil=self._osutil, config=self._config,
+            executor=self._executor)
+        return upload_submitter(call_args)
+
+    def _validate_subscribers(self, subscribers):
```

Comment:
Another benefit of creating a base class is that you can move all this validation to the metaclass so we don't have to validate the object at upload time.  A given class will either validate or not validate, and if a user wants to reuse the same subscriber instance each time we're unnecessarily validating the class repeatedly.